### PR TITLE
Add support for getting and setting multiple HTTP fields.

### PR DESCRIPTION
**Pull request by binary1248**
Implements #134.

### CODE REVIEW

File: src/SFML/Network/Http.cpp [hunk 1]
```diff
@@ -61,7 +61,20 @@ Http::Request::Request(const std::string& uri, Method method, const std::string&
 ////////////////////////////////////////////////////////////
 void Http::Request::setField(const std::string& field, const std::string& value)
 {
-    m_fields[toLower(field)] = value;
+    m_fields.erase(toLower(field));
+    m_fields.insert(std::pair<std::string, std::string>(toLower(field), value));
```
Comment (LaurentGomila) [inline src/SFML/Network/Http.cpp]: ```std::make_pair```?

**Comment by LaurentGomila** [comment]
Wouldn't a ```std::map<std::string, std::vector<std::string> >``` simplify the implementation?

**Comment by MarioLiebisch** [comment]
I'd scrap the whole thing and do it differently.

To cite [RFC2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2):

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.

So we could safely combine all these into one value when receiving headers. In addition, provide a new `sf::Http::Request::appendField(field, value)`, which will essentially append `value` to an existing field or create a new one.

**Comment by select-this** [comment]
> How to escape embedded commas? Is that defined somewhere? Add quotes? But how to escape quotes then?

I haven't been able to find any literature on the subject, but I'm surmising from the wording of the RFC that anything that contains a comma in the value excludes it from being able to be combined into one field (e.g. I've seen the Date field include a comma in the date format, and therefore there can only be one Date field in the HTTP message). As stated, this is a guess based on examining a few requests and reading between the lines of the RFC, I have no evidence for this yet.

**Comment by binary1248** [comment]
> Wouldn't a `std::map<std::string, std::vector<std::string> >` simplify the implementation?

No, because you would have to iterate over 2 containers instead of one. `std::multimap` takes care of grouping pairs with identical keys together so we don't have to do anything special when constructing the header.

> So we could safely combine all these into one value when receiving headers. In addition, provide a new `sf::Http::Request::appendField(field, value)`, which will essentially append value to an existing field or create a new one.

Why multiple calls to `setField()` and `appendField()` when you could do the same with one call and a container? The multiple fields could be cached for future use somewhere, and your suggestion would force the user to iterate over their container each time a request has to be constructed.

Also, it is nicer of us to provide the split data to the user if we already got it split. There is no point concatenating the multiple lines into a single string just so that the user can pick it apart again. The RFC says that combining into a single list is allowed, but does not force one to do so. If the user decides to combine themselves, then they can already do so with the current API. If the server combines, the current API also takes care of that. What the current API doesn't do is provide a possibility to send multiple lines with the same field name or parse server responses with such a payload. It is not up to us to ask why one would need such a feature, there is enough broken software out there that could warrant having to support such things.

> How to escape embedded commas? Is that defined somewhere? Add quotes? But how to escape quotes then?

You URL encode them. Many cookies are also stored as base-64 keys. I don't think it is common to send text strings containing punctuation in the message headers. After all, they were never meant to be consumed by a human anyway.

**Comment by select-this** [comment]
> I don't think it is common to send text strings containing punctuation in the message headers.

Anything pertaining to dates appears to violate this assumption. 

Take this request example from http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html:

```
HTTP/1.1 206 Partial content
Date: Wed, 15 Nov 1995 06:25:24 GMT
Last-Modified: Wed, 15 Nov 1995 04:58:08 GMT
Content-Range: bytes 21010-47021/47022
Content-Length: 26012
Content-Type: image/gif
```

Both the Date and Last-Modified fields have commas in them, not that you'd ever need to send multiples of these fields.

**Comment by LaurentGomila** [comment]
> No, because you would have to iterate over 2 containers instead of one

That would just add one line to the ```prepare()``` function, but simplify others and avoid a full copy of the field array in ```getFieldsValues```.

```cpp
void Http::Request::setField(const std::string& field, const std::string& value)
{
    std::vector<std::string> values;
    values.push_back(value);
    setFieldValues(field, {value});

    // C++11: setFieldValues(field, {value});
}

void Http::Request::setFieldValues(const std::string& field, const std::vector<std::string>& values)
{
    m_fields[toLower(field)] = values;
}

const std::vector<std::string>& Http::Response::getFieldValues(const std::string& field) const
{
    static const std::vector<std::string> empty;
    xyz.::iterator it = m_fields.find(toLower(field));
    return it != m_fields.end() ? it->second : empty;
}

void Http::Response::parseFields(std::istream &in)
{
    ...

    m_fields[toLower(field)].push_back(value);
}
```

**Comment by binary1248** [comment]
Fixed.

Also, after a bit more reading, I don't think handling HTTP message headers is as simple as we think. The RFC states that multiple independent lines with the same header name can be combined using comma separated lists, but this leads to ambiguity if the comma is already used as part of the header value (they know this). Cookies for example use the semicolon as a separator in their value because the expiry dates can contain commas (which don't have to be escaped or part of quoted strings).

Anyone who wants to write a really hardened HTTP client/server will have to take care of all the edge cases themselves when it comes to header parsing. All SFML can do is provide each line as it is and let the user deal with the rest.

File: include/SFML/Network/Http.hpp [hunk 2]
```diff
@@ -257,6 +273,20 @@ class SFML_NETWORK_API Http : NonCopyable
...
+        /// If the field \a field is not found in the response header,
+        /// an empty container is returned. This function uses
+        /// case-insensitive comparisons.
+        ///
+        /// \param field Name of the field whose values to get
+        ///
+        /// \return Values of the field, or empty container if not found
+        ///
+        ////////////////////////////////////////////////////////////
+        std::vector<std::string> getFieldValues(const std::string& field) const;
```
Comment (LaurentGomila) [inline include/SFML/Network/Http.hpp]: Should be ```const std::vector<std::string>&```

**Comment by binary1248** [comment]
Fixed.

**Comment by binary1248** [comment]
I've been thinking... maybe it's a better idea to just scrap this approach and simply provide access to the raw HTTP header content through a `getHeader()` method, similar to `getBody()`. That way, the user can perform all of the processing necessary to get their job done in their own code. If their use case gets really complex, they can think of a solution to handle it on their own instead of relying on us to properly parse/split it for them.

**Comment by LaurentGomila** [comment]
So why don't we remove ```sf::Http```? Users can still build a better HTTP solution around ```sf::TcpSocket``` themselves, after all.

**Comment by binary1248** [comment]
I figured that was what we were going to do in SFML 3 anyway... :stuck_out_tongue: The question is whether we bother with this issue for the time being or not.

**Comment by eXpl0it3r** [comment]
So what's the verdict here?

**Comment by binary1248** [comment]
Bump.

**Comment by binary1248** [comment]
Bump.

**Comment by TankOs** [comment]
It's not like we are the first ones who implement HTTP. We should take a look at source code, e.g. nginx.

**Comment by binary1248** [comment]
The "problem" is that HTTP isn't as simple as it seems at first glance. Right now, SFML implements "just enough" to get rudimentary HTTP 1.0 (not even 1.1) working. The question is where does SFML draw the line? Obviously it should not become a fully featured HTTP library. If the user needs more HTTP functionality, SFML should just provide them with enough on which they can build upon.

**Comment by TankOs** [comment]
ACK. :accept:

Even in web-centric frameworks I still set headers etc. myself, including delimiters. Basic HTTP support in SFML is great, but if you want to do more than "GET /some/uri" or post some data, SFML's HTTP class should act like a simple utility class IMHO.

And to be honest, I'd be annoyed if sf::Http tried to wrap everything of the protocol, instead of doing a simple setHeader("Accept", "blahblah");

**Comment by eXpl0it3r** [comment]
That still leaves the question open, what's the verdict? :wink:

**Comment by Bromeon** [comment]
> binary1248:
If the user decides to combine [header fields] themselves, then they can already do so with the current API. If the server combines, the current API also takes care of that. What the current API doesn't do is provide a possibility to send multiple lines with the same field name or parse server responses with such a payload.

I think that's the most important observation here. The current API is limited. We should not provide all the convenience functions that sophisticated HTTP clients do, but we should allow the user to build something more powerful with basic tools.

Regarding whether this PR (`getFieldValues()` + `setFieldValues()`) or a more low-level approach (`getHeader()` + `setHeader()`?) is appropriate: can you imagine other cases where setting and retrieving single header fields is *still* too limited? I currently only see the number of HTTP methods as a potential limitation, but that's not part of the header...

**Comment by binary1248** [comment]
The current implementation is limited in other ways as well. SFML has to declare itself as only supporting HTTP/1.0 because as soon as the server uses any mechanisms that HTTP/1.1 introduces, half of the assumptions made in the implementation are no longer valid. As an example, the fact that HTTP/1.1 servers might not immediately close the connection so that it can be reused for further requests means that SFML would wait until it times out (which can take a while) before proceeding with message parsing. HTTP/1.0 mandated that the server terminates the connection after servicing a request.

As for your question, there aren't really any other cases. Like I already said, HTTP headers are "free form" to a certain extent. Most of the time, they are trivial to parse and provide through an API like SFML's, but in more complex cases, we shouldn't make any assumptions and leave the complicated parsing procedure up to the user if they reckon they can deal with it (I assume that they already have a general idea of what the header would look like). We are really only talking about the headers here, and nothing else. Either they fit into the trivial form, or they don't. We currently don't properly handle (or provide a way to handle) the latter case.

**Comment by Ostefanini** [comment]
Okay, so you will remove sf::http for the next version ? I don't know if you see it seems like c++17 will support networking..
https://meetingcpp.com/index.php/br/items/c17-and-its-technical-specifications.html

**Comment by eXpl0it3r** [comment]
Looks like networking isn't going to make C++17.

As I see it, we could proceed with this PR and it won't really change the status of SFML's lack of proper/useful HTTP support, but it would allow cases as in #134.

For how we could fix SFML's HTTP support we should start a discussion in the SFML Development section and see what options exist.

If the concern is API cluttering, then we could also have the discussion first, but it needs to happen and this PR and issue will be removed for the 2.5 milestone.

**Comment by vfjpl** [comment]
Bump.

What's the status on this?

**Comment by Bromeon** [comment]
I think this might need a general design discussion w.r.t. **should SFML 3 still include HTTP support?**
Reasons I can think of to remove `sf::Http` from SFML 3:

* There are other libraries specialized at this task, with much more feature-rich APIs
* HTTPS is not supported. With encrypted websites being much more the norm than they were in the 2000s, an unencrypted HTTP connection might need special handling (e.g. downgrading) server-side, and is unlikely to be integrated in web APIs. I'm also not sure if we should still encourage HTTP in 2020.
* The main use case for this was probably high-score tables for a simple game. I don't know how many people effectively used this feature.

**Comment by binary1248** [comment]
I was always for removing HTTP support since less and less servers actually offer unencrypted service these days with things like letsencrypt.

The only argument I can think of for non-encrypted HTTP would be site-local communication between e.g. machines, but even in that scenario there are much better ways of transferring data than via HTTP. HTTP really is a web technology, and insecure web technology is basically obsolete.

**Comment by binary1248** [comment]
I think by now, the C++ ecosystem has progressed to the point that if this functionality is needed there are better more powerful libraries out there that can be used to solve this problem and more.

**Pull request closed without merging**